Add Dockerfile to containerize the Spring Boot application

tiago-paiva opened the pull request:
- Uses OpenJDK 21
- Copies the built JAR file into the container
- Exposes port 8080 and sets up the entrypoint

tiago-paiva committed:
Add Dockerfile to containerize the Spring Boot application
- Uses OpenJDK 21
- Copies the built JAR file into the container
- Exposes port 8080 and sets up the entrypoint

tiagomemt committed:
Adicionado multi-stage build no Dockerfile para compilar o aplicativo com Maven antes de criar a imagem final

tiago-paiva committed:
Substitui H2 por PostgreSQL e configura conexão do PostgreSQL via Docker Compose

luisguilermes commented on docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+
+  app:
+    build: .
+    container_name: tm-ctn
```
Sugestão: Separe os arquivos de configuração do Docker Compose em docker-compose.yaml e docker-compose.dev.yaml.

No docker-compose.yaml, configure apenas o serviço do PostgreSQL, permitindo que você desenvolva e teste localmente sem precisar subir a aplicação inteira.

Já o docker-compose.dev.yaml deve incluir o docker-compose.yaml usando a diretiva include, garantindo que, além do banco de dados, a aplicação também seja iniciada. Dessa forma, qualquer pessoa pode testar a aplicação completa usando apenas o Docker, facilitando o onboarding e a consistência no ambiente de desenvolvimento.

luisguilermes commented on src/main/resources/application.properties:
```diff
@@ -1,7 +1,7 @@
-spring.datasource.password=
-spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
-spring.h2.console.enabled=true
+spring.datasource.url=jdbc:postgresql://postgres:5432/taskmanager
```
Sugestão: Prefira utilizar o formato YAML(application.yaml) no application.properties, pois ele é mais legível e organizado.

Além disso, separe as configurações específicas de cada ambiente em arquivos distintos. Por exemplo, a URL do banco de dados e outras variáveis sensíveis devem estar em arquivos específicos, como application-local.yaml para o ambiente local e application-dev.yaml para o ambiente de desenvolvimento.

Essa abordagem facilita a manutenção, evita conflitos de configuração e garante que cada ambiente tenha suas próprias definições de maneira clara e segura.

luisguilermes commented on docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+    depends_on:
+      - postgres
+    environment:
+      SPRING_DATASOURCE_URL: jdbc:postgresql://postgres:5432/taskmanager
```
**Pergunta:** Você chegou a testar a aplicação localmente, sem rodar no docker, integrando e salvando os dados no PostgreSQL?

É bem provável que, se eu rodar a aplicação localmente, ela não funcione corretamente.

luisguilermes commented:
**Dica:** crie uma documentação no readme do projeto mostrando como executar o projeto!

tiago-paiva commented:
Sim, eu testei a aplicação localmente integrando com o PostgreSQL no Docker antes de configurar o Docker Compose. No entanto, após a configuração do Compose, precisei ajustar o application.properties para compatibilidade.

Vou seguir suas sugestões sobre a organização das configurações por ambiente e a utilização do formato YAML para melhorar a manutenção e evitar conflitos.

tiago-paiva committed:
substitui o arquivo application.properties pelo application.yaml para facilitar a leitura e organização

tiago-paiva committed:
Adicionado .env para centralizar configurações do ambiente.
Atualizado docker-compose.yml e application.yaml para usar variáveis de ambiente.
